aesd-char-driver/main.c

```diff
@@ -0,0 +1,262 @@
+ *
+ */
+
+#include <linux/module.h>
+#include <linux/init.h>
+#include <linux/printk.h>
+#include <linux/types.h>
+#include <linux/cdev.h>
+#include <linux/fs.h>   // file_operations
+#include <linux/string.h>
+#include <linux/slab.h>
+
+#include "aesdchar.h"
+int aesd_major = 0; // use dynamic major
+int aesd_minor = 0;
+
+MODULE_AUTHOR("Harsh Beriwal"); /** TODO: fill in your name **/
+MODULE_LICENSE("Dual BSD/GPL");
+
+struct aesd_dev aesd_device;
+
+int aesd_open(struct inode *inode, struct file *filp)
+{
+    struct aesd_dev *a_dev;
+    PDEBUG("open");
+
+    a_dev = container_of(inode->i_cdev, struct aesd_dev, cdev);
+    filp->private_data = a_dev;
+    return 0;
+}
+
+int aesd_release(struct inode *inode, struct file *filp)
+{
+    struct aesd_dev *a_dev;
+    PDEBUG("release");
+    
+    a_dev = container_of(inode ->i_cdev, struct aesd_dev,cdev);
+    (void)a_dev;
+    //filp->private_data = NULL;
+    return 0;
+}
+
+ssize_t aesd_read(struct file *filp, char __user *buf, size_t count,
+                  loff_t *f_pos)
+{
+    struct aesd_buffer_entry *element;
+    struct aesd_dev *a_dev;
+    ssize_t retval = 0;
+    int buffer_count = 0;
+    size_t offset;
+    
+    PDEBUG("read %zu bytes with offset %lld", count, *f_pos);
+
+    a_dev = (struct aesd_dev*)filp->private_data;
+
+    if (mutex_lock_interruptible(&a_dev->lock)!=0) {
+	PDEBUG(KERN_ERR "Couldn't acquire Mutex\n");
+	 mutex_unlock(&a_dev->lock);
+	 return retval;
+    }
+
+    element = aesd_circular_buffer_find_entry_offset_for_fpos(&a_dev->circularBuffer, *f_pos, &offset);
+    if(element==NULL) {
+    	 mutex_unlock(&a_dev->lock);
+	 return retval;
+    }
+
+    if ((element->size - offset) < count)  {
+        *f_pos = *f_pos + (element->size - offset);
+        buffer_count = element->size - offset;
+    } else  {
+        *f_pos = *f_pos + count;
+        buffer_count = count;
+    }
+
+    if (copy_to_user(buf, element->buffptr+offset, buffer_count)) 
+    {
+	retval = -EFAULT;
+	mutex_unlock(&a_dev->lock);
+	return retval;
+    }
+
+    retval = buffer_count;
+
+    mutex_unlock(&a_dev->lock);
+
+    return retval;
+}
+
+ssize_t aesd_write(struct file *filp, const char __user *buf, size_t count,
+                   loff_t *f_pos)
+{
+    ssize_t retval = 0;
+    int bytes_to_send = 0, index =0; 
+    int entries = 0, total_bytes = 0; 
+    char *element;
+    const char *freed_buff;
+    struct aesd_buffer_entry w_buffer;
+    struct aesd_dev *a_dev = filp->private_data;
+
+    PDEBUG("write %zu bytes with offset %lld", count, *f_pos);
+    
+    if (mutex_lock_interruptible(&a_dev->lock)!=0) {
+	PDEBUG(KERN_ERR "Couldn't acquire Mutex\n");
+	return -EFAULT;
+    }
+    element = (char *)kmalloc(count, GFP_KERNEL);
+    if (element == NULL)  {
+        retval = -ENOMEM;
+        mutex_unlock(&a_dev->lock);
+  	return retval;
+    }
+    if (copy_from_user(element, buf, count))  {
+        retval = -EFAULT;
+	mutex_unlock(&a_dev->lock);
```

Comment:
@harshberiwal what happens to the memory allocated for `element` in this code path?